DOCSP-30562: operation error handling (#35)
mongodb/docs-rust: source/op-error-handling.txt

.. _rust-operation-errors:

========================
Operation Error Handling
========================

.. contents:: On this page
   :local:
   :backlinks: none
   :depth: 2
   :class: singlecol

Overview
--------

This page describes errors you might encounter when
using the {+driver-long+} to perform MongoDB operations.

.. note::

>    This page addresses only operation error handling. If you encounter

Reviewer: ```suggestion
   This page only addresses operation error handling. If you encounter
```

Author: Per the[ style guide](https://www.mongodb.com/docs/meta/style-guide/terminology/alphabetical-terms/#o:~:text=Don%27t%20hyphenate.-,only,-Place%20only%20preceding), the only is currently in the correct place

Reviewer: Oh good to know, sorry about that!

   any other issues with MongoDB or the driver, visit the following
   resources:

   - The :ref:`Issues & Help <rust-issues-and-help>` page, which has
     information about reporting bugs, contributing to the driver, and 
     finding more resources
   - The `MongoDB Community Forums <https://community.mongodb.com>`__ for
     questions, discussions, or general technical support

   ..
     - The :ref:`Frequently Asked Questions (FAQ) <golang-faq>` for the
       {+driver-short+}
     - :ref:`rust-connection-troubleshooting`

Error Type
----------

> Most of the exceptions that the driver throws when executing operations

Reviewer: S: style guide prefers "generate" to "throw" (in the "Write for a Global Audience" section)

are of the `Error <{+api+}/error/struct.Error.html>`__ type. The
``Error`` instance that an operation
returns contains the ``kind`` field, which describes the type of error
that occurred. The ``kind`` field has a value of type `ErrorKind
<{+api+}/error/enum.ErrorKind.html>`__, which has many variants for

> different kinds of errors.

Reviewer: S: it might be worthwhile to include a list of some of these variants, just because "many variants for different kinds of errors" seems pretty vague

For example, if you attempt to perform an insert operation that
duplicates an ``_id`` field value that is already in the collection,
the driver returns an ``Error`` instance and prints the following error
message:

.. code-block:: none
   :copyable: false

   Error: Error { kind: Write(WriteError(WriteError { code: 11000,
   code_name: None, message: "E11000 duplicate key error collection:
   db.test_coll index: _id_ dup key: { _id: 1 }", details: None })), labels:
   {}, wire_version: None, source: None }

In the preceding error message, the value of the ``kind`` field is
``Write``, which applies when an error occurs when

> you try to execute a write operation.

Reviewer: S:

1. This sentence is a little wordy and repeats "when," you could change to something like:
"... the value of the ``kind`` field is ``Write``. This value applies when an error occurs during a write operation."

2. style guide is weird about "execute," could replace with "run" (or rephrased like the suggestion above)

Author: reworded

Connection Errors
-----------------

You might successfully connect to a MongoDB server, but the connection

> pool might be cleared due to a concurrent operation error. In this

Reviewer: S: instead of using "might" as a helping verb twice, you can reword this to make the verbs a little simpler:
"A concurrent operation error might interrupt a successful connection to a MongoDB server and clear the connection pool"

situation, the driver raises an ``Error`` type where the value of the
``kind`` field is ``ConnectionPoolCleared``. The error message describes
the reason that the concurrent operation failed.

Depending on the circumstances that produce the error, the driver may
add a ``RetryableWriteError`` label to the error, as shown in the
following error message:

.. code-block:: none
   :copyable: false

   Error { kind: ConnectionPoolCleared { message: "Connection pool for
   localhost:27017 cleared because another operation failed with: Kind:
   I/O error: timed out, labels: {}" }, labels: {"RetryableWriteError"},
   wire_version: None, source: None }

This label indicates that the error is the error is write-retryable.

Write Error Types
-----------------

When the driver experiences an error while performing a write operation,
it raises an error where the ``kind`` field value is ``Write``.
The body of the ``Write`` variant is the enum `WriteFailure
<{+api+}/error/enum.WriteFailure.html>`__, which
takes a value of type `WriteError <{+api+}/error/struct.WriteError.html>`__ or
`WriteConcernError <{+api+}/error/struct.WriteConcernError.html>`__.

Write Concern Error
~~~~~~~~~~~~~~~~~~~

> The driver raises a ``WriteConcernError`` error when you execute a write

Reviewer: ```suggestion
The driver raises a ``WriteConcernError`` error when you run a write
```

Author: changed all to perform

operation and the driver cannot satisfy the specified write concern. For
example, if you specify a write concern of ``majority`` for

> operations on a replica set with three nodes, the driver would return

Reviewer: ```suggestion
operations on a replica set with three nodes, the driver returns
```

For example, the driver raises a ``WriteError`` error if you attempt to
insert a document into a collection that violates the collection's
schema validation rules. Suppose the collection has a rule where the
value of the ``quantity`` field must be an ``int`` type. If you
attempt to insert a document where the value of ``quantity`` is
``"three"``, the driver prints the following error message:

.. code-block:: none
   :copyable: false

   Error: Error { kind: Write(WriteError(WriteError { code: 121, code_name:
   None, message: "Document failed validation", details:
   Some(Document({"failingDocumentId": Int32(1), "details":
   Document({"operatorName": String("$jsonSchema"), "title":
   String("Numerical Validation"), "schemaRulesNotSatisfied":
   Array(...)})})) })), labels: {},
   wire_version: None, source: None }

In the preceding error message, the ``message`` field describes the
reason for the error, and the ``details`` field provides specific
details about the failing operation.

Additional Information
----------------------

To learn more about write concerns, see :manual:`Write Concern
</reference/write-concern/>` in the Server manual.

>      

Reviewer: S: you link to some API documentation on this page (Error, ErrorKind, etc) that could be good to add here

Reviewer: actually I'm not sure if we usually include API links in a separate section that are also linked in the text, so disregard if not

Author: For this kind of page (not a fundamentals page), I think linking inline could be helpful. What do you also think about removing the additional information section and moving the content that's currently there to the places where they are more relevant?

Reviewer: That sounds good to me!

.. _rust-operation-errors:

========================
Operation Error Handling
========================

.. contents:: On this page
   :local:
   :backlinks: none
   :depth: 2
   :class: singlecol

Overview
--------

This page describes errors you might encounter when
using the {+driver-long+} to perform MongoDB operations. Once you
understand the types of operation errors that the driver raises, you can take
appropriate actions to either handle them or correct the error-causing code.

.. note::

   This page addresses only operation error handling. If you encounter
   any other issues with MongoDB or the driver, visit the following
   resources:

   - The :ref:`Issues & Help <rust-issues-and-help>` page, which has
     information about reporting bugs, contributing to the driver, and 
     finding more resources
   - The `MongoDB Community Forums <https://community.mongodb.com>`__ for
     questions, discussions, or general technical support

   ..
     - The :ref:`Frequently Asked Questions (FAQ) <golang-faq>` for the
       {+driver-short+}
     - :ref:`rust-connection-troubleshooting`

Error Type
----------

> Most of the exceptions that the driver generates when performing operations
> are of the `Error <{+api+}/error/struct.Error.html>`__ type. The

Reviewer 2: ```suggestion
Errors returned by the driver when performing operations are of the `Error <{+api+}/error/struct.Error.html>`__ type. The
```

[`Result` documents](https://github.com/mongodb/mongo-rust-driver/blob/6a9ef26bd1adc721fb59e61a6895d95cef3e3200/src/error.rs#L37):
> The result type for all methods that can return an error in the `mongodb` crate.

I expect all errors are returned as `Error`.

Suggest avoiding the word "exception", since Rust does not have exceptions.

   Error: Error { kind: Write(WriteError(WriteError { code: 11000,
   code_name: None, message: "E11000 duplicate key error collection:
   db.test_coll index: _id_ dup key: { _id: 1 }", details: None })), labels:
   {}, wire_version: None, source: None }

In the preceding error message, the value of the ``kind`` field is
``Write``. To learn more about this type of error, see the :ref:`Write
Error Types <rust-error-write-errors>` section of this guide.

Connection Errors
-----------------

A concurrent operation error might clear the connection pool,
interrupting your connection to the server. In this
situation, the driver raises an ``Error`` type where the value of the
``kind`` field is ``ConnectionPoolCleared``. The error message describes
the reason that the concurrent operation failed.

To learn how to adjust your connection pool to address this error, see
:manual:`Tuning Your Connection Pool Settings
</tutorial/connection-pool-performance-tuning/>` in the Server manual.

Depending on the circumstances that produce the error, the driver may
add a ``RetryableWriteError`` label to the error, as shown in the
following error message:

.. code-block:: none
   :copyable: false

>    :emphasize-lines: 1

Reviewer: S: I'm not sure if this first line needs to be highlighted -- since you describe the RetryableWriteError label before and after the code block, maybe you could highlight line 3 instead (or remove the highlighting). Not a necessary change though if you want to emphasize the kind field
